src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -267,7 +276,12 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
                 }
             };
             debug!("store to var {:?}", index);
-            self.local_qualif[index] = Some(self.qualif);
+            match &mut self.local_qualif[index] {
+                // update
+                Some(ref mut qualif) => *qualif = *qualif | self.qualif,
```

Comment:
You should only do this if the place is a projection, otherwise you *are* replacing the value.

Reply:
That would mean there's a difference between

```rust
const FOO: &Option<Cell<usize>> = {
    let mut a = (Some(Cell::new(0)),);
    a.0 = None; // resets `qualif(a)` to `qualif(None)`
    &{a.0}
};
```

and 

```rust
const FOO: &Option<Cell<usize>> = {
    let mut a = Some(Cell::new(0));
    a = None; // resets `qualif(a)` to `qualif(None)`
    &{a}
};
```

Reply:
I guess you want to be conservative? r=me with a comment that explains that we could allow more for whole-field assignments but we're being uniform/conservative/etc.

Reply:
That doesn't seem worth it though, just to support such obscure code?

Reply:
Well... "just this" isn't worth it. But I assume we'll want it anyway at some point when we have loops and everything. Any static analysis will be severely outgunned by an analysis on the final value.